src/Framework/AspNetCoreAnalyzers/src/SourceGenerators/PublicTopLevelProgramGenerator.cs

```diff
@@ -0,0 +1,49 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace Microsoft.AspNetCore.SourceGenerators;
+
+[Generator]
+public class PublicProgramSourceGenerator : IIncrementalGenerator
+{
+    public void Initialize(IncrementalGeneratorInitializationContext context)
+    {
+        var internalGeneratedProgramClass = context.CompilationProvider.Select((compilation, cancellationToken) =>
+        {
+            var program = compilation.GetTypeByMetadataName("Program");
+            // If the program class is already public, we don't need to generate anything.
+            if (program is null || program.DeclaredAccessibility == Accessibility.Public)
+            {
+                return null;
+            }
+            if (program.DeclaringSyntaxReferences.Length > 1)
+            {
+                return null;
+            }
+            // If the `Program` class is already declared in user code, we don't need to generate anything.
+            if (program.DeclaringSyntaxReferences.SingleOrDefault()?.GetSyntax(cancellationToken) is ClassDeclarationSyntax)
```

Comment:
Remind me whether `ClassDeclarationSyntax` covers structs?

Reply:
It looks like there's a separate [StructDeclarationSyntax](https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.csharp.syntax.structdeclarationsyntax?view=roslyn-dotnet-4.9.0) for that.

Reply:
Do we need to accept that too?  I'm pretty sure you can `struct Program { public static void Main() {} }`.

Reply:
Yep, although in this case, we can use a `program.TypeKind == TypeKind.Struct` check since we know the only way a `Program` type will be a `struct` is it if has already been defined in source code so the `program.DeclaringSyntaxReference` isn't necessary.